code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -73,36 +113,173 @@ public void computeInstanceActivateReply(Event event) {
         }
     }
 
-    protected boolean handleUnreportedKnownInstances(long agentId, Map<String, String> knownInstances,
-            ReportedInstances reportedInstances, boolean checkOnly) {
-
-        if (knownInstances != null) {
-            for (Map.Entry<String, String> knownInstance : knownInstances.entrySet()) {
-                if(!reportedInstances.byUuid.contains(knownInstance.getKey())
-                        && !reportedInstances.byExternalId.contains(knownInstance.getValue())) {
-                    if (checkOnly) {
-                        return true;
-                    } else {
-                        // If this known instance was not reported, schedule a potential restart.
-                        restart(knownInstance.getKey());
-                    }
+    /*
+     * If checkOnly is true, will raise a ContainersOutOfSync exception, indicating this should be reran with checkOnly set to false.
+     */
+    void syncContainers(Map<String, Instance> knownInstances, ReportedInstances reportedInstances, long agentAccountId, long agentId, long hostId,
+            boolean checkOnly) {
+        Map<String, ReportedInstance> needsSynced = new HashMap<String, ReportedInstance>();
+        Map<String, String> syncActions = new HashMap<String, String>();
+        Set<String> needsHaRestart = new HashSet<String>();
+
+        determineSyncActions(knownInstances, reportedInstances, needsSynced, syncActions, needsHaRestart, checkOnly);
+
+        for (String uuid : needsHaRestart) {
+            restart(uuid);
+        }
+
+        for (Map.Entry<String, ReportedInstance> syncEntry : needsSynced.entrySet()) {
+            ReportedInstance ri = syncEntry.getValue();
+            String syncAction = syncActions.get(syncEntry.getKey());
+            if (EVENT_INSTANCE_FORCE_STOP.equals(syncAction)) {
+                forceStop(ri.getExternalId(), agentId);
+            } else if (HA_RESTART.equals(syncAction)) {
+                String uuid = ri.getInstance() != null ? ri.getInstance().getUuid() : ri.getUuid();
+                restart(uuid);
+            } else {
+                scheduleContainerEvent(agentAccountId, hostId, ri, syncAction);
+            }
+        }
+    }
+
+    void determineSyncActions(Map<String, Instance> knownInstances, ReportedInstances reportedInstances, Map<String, ReportedInstance> needsSynced,
+            Map<String, String> syncActions, Set<String> needsHaRestart, boolean checkOnly) {
+        Map<String, Instance> inRancher = new HashMap<String, Instance>(knownInstances);
+        Map<String, ReportedInstance> onHost = new HashMap<String, ReportedInstance>(reportedInstances.byExternalId);
+        for (Map.Entry<String, ReportedInstance> reported : reportedInstances.byUuid.entrySet()) {
+            Instance ki = knownInstances.get(reported.getKey());
+            if (ki != null) {
+                onHost.remove(reported.getValue().getExternalId());
```

Comment:
Minor, but lines 152-155 seem to be identical to 169-172. May be it would make sense to extract it to some helper method.